File: heap/heap.rb
require 'pry'


# Heap Priority Queu
class Heap
  attr_accessor :pq, :n

  def initialize
    @pq = [] # heap order complete binary tree
    @n = 0 # in pq[1..N] with pq[0] unused
  end

  def empty?
    @n == 0
  end

  def size
    @n
  end

  def insert(v)
    @pq[@n += 1] = v
    swim(@n) # position in heap
  end

  # trickleup
  def swim(k)
    while k > 1 && less((k / 2).floor, k) # parent: k/2 parent less than child?
      swap((k/2).floor, k)
      k = k / 2
    end
  end

  def del_max
    max = @pq[1]  # Retrieve max key from top.
    swap(1, @n) # Exchange with last item.
    @n -= 1
    @pq[@n + 1] = nil # Avoid loitering. max added to end on swap make nil
    sink(1) # Restore HEAP property
    max
  end

  # Trickledown
  def sink(k)
    while 2 * k <= @n
      j = 2 * k # child

      if !@pq[j + 1].nil? # check if there is a right child
        j += 1 if j > 1 && @pq[j] < @pq[j + 1] # if left child less than right child
      end
      break if @pq[k] > @pq[j] # If parent greater than child break

      swap(k, j)
      k = j
    end
  end

  def less(i, j)
    @pq[i] < @pq[j]
  end

  def swap(i, j)
    temp = @pq[i]
    @pq[i] = @pq[j]
    @pq[j] = temp
  end
end


heap= Heap.new
# heap.insert(4)
# heap.insert(1)
# heap.insert(2)
# heap.insert(1)
# SORTEXAMPLE
heap.insert('S')
heap.insert('O')
heap.insert('R')
heap.insert('T')
heap.insert('E')
heap.insert('X')
heap.insert('A')
heap.insert('M')
heap.insert('P')
heap.insert('L')
heap.insert('E')


binding.pry
puts "cookie"

File: heap/heap.js

class Heap {
  constructor() {
    this.n = 0;
    this.pq = [];
  }

  size() {
    return this.n;
  }

  isEmpty() {
    this.n === 0;
  }

  swim(k) {
    while(k > 1 && this.less(Math.floor(k / 2), k)){
      this.exch(Math.floor(k / 2), k);
      k = Math.floor(k / 2)
    }
  }

  sink(k) {
    while(2 * k <= this.n) {
      let j = 2 * k;

      if(this.pq[j + 1] != null) {
        if(k > 1 && this.less(j, j + 1)){
          j++;
        }
      }

      if(this.pq[k] > this.pq[j]) {
        break;
      }
      this.exch(k, j)
      k = j
    }
  }

  insert(v) {
    this.pq[++this.n] = v;
    this.swim(this.n);
  }

  delMax() {
    let max = this.pq[1];
    this.exch(1, this.n--);
    this.pq[this.n + 1] = null;
    this.sink(1);
    return max;
  }

  less(i, j) {
    return this.pq[i] < this.pq[j];
  }

  exch(i, j) {
    let temp = this.pq[i];
    this.pq[i] = this.pq[j];
    this.pq[j] = temp;
  }
}

let heap = new Heap();
heap.insert("t")
heap.insert("p")
heap.insert("r")
heap.insert("n")
heap.insert("h")
heap.insert("o")
heap.insert("a")
heap.insert("e")
heap.insert("i")
heap.insert("g")
heap.insert("s")

console.log(heap.pq)
console.log(heap.isEmpty())
console.log(heap.size())

console.log('----cookie-----')
heap.pq = [null, 't', 's', 'r', 'n', 'p','o', 'a', 'e', 'i', 'g', 'h']
heap.delMax()

console.log(heap.pq)
heap.delMax()
console.log(heap.pq)

